### PR TITLE
Add a livetime map on dataset

### PR DESCRIPTION
**Pull request by AtreyeeS**
This addresses #3903 by adding a new member on the `MapDataset`.
The serialisation is not implemented.

### CODE REVIEW

File: gammapy/datasets/map.py [hunk 1]
```diff
@@ -204,6 +208,7 @@ def __init__(
         gti=None,
         meta_table=None,
         name=None,
+        livetime_map=None,
```
Comment (adonath) [inline gammapy/datasets/map.py:211]: Please rename to `livetime`, the other attributes do not carry the `_map` prefix either. The type can be clarified in the docstring.

File: gammapy/datasets/map.py [hunk 2]
```diff
@@ -573,6 +579,9 @@ def from_geoms(
             "gti", GTI.create([] * u.s, [] * u.s, reference_time=reference_time)
         )
         kwargs["mask_safe"] = Map.from_geom(geom, unit="", dtype=bool)
+        kwargs["livetime_map"] = kwargs["mask_safe"].reduce_over_axes(
```
Comment (adonath) [inline gammapy/datasets/map.py:582]: The livetime can be initialized as a non-masked map. Just as for the other qualities the mask will be applied in `.stack()` and `.to_masked()`

File: gammapy/makers/map.py [hunk 3]
```diff
@@ -412,4 +414,9 @@ def run(self, dataset, observation):
 
             kwargs["edisp"] = edisp
 
+        kwargs["livetime_map"] = (
+            kwargs["mask_safe"].reduce_over_axes(func=np.logical_or)
```
Comment (adonath) [inline gammapy/makers/map.py:418]: I would leave the livetime map un-masked here...

**Review by adonath — COMMENTED** [review]
Thanks @AtreyeeS, this is useful information added to the `MapDataset`. One more general comment: I think the livetime map also has to be stored as 3D map. When stacking different observations can have different energy thresholds resulting in an energy dependent livetime.

Currently the scalar livetime is stored in the exposure map as well, this could probably be deprecated. But I would need to think about how...

**Comment by AtreyeeS** [comment]
@adonath We discussed this a bit yesterday, and if we correct for the energy, why not for the acceptance? This information is then correctly available in the exposure map, and, we show how to calculate a livetime map in the HESS tutorial.

This map is useful to because people often want to quote "there were x hours of observations of this region"

I added the `_map` tag to not confuse with the `livetime` present on the `info_dict` (which is taken from the  exposure map). Even if deprecated, we need to support both for the next version, no?

File: gammapy/datasets/map.py (hunk 2 above)
Comment (adonath) [reply gammapy/datasets/map.py:582]: No, but `Map.from_geom(geom, unit="s")` is simpler I guess..

File: gammapy/makers/map.py (hunk 3 above)
Comment (AtreyeeS) [reply gammapy/makers/map.py:418]: The livetime map for a single observation should take the mask into account

Comment (adonath) [reply gammapy/makers/map.py:418]: For single observations we leave the quantities unmasked. The mask is only applied when stacking or calling `.to_masked()`. This has the huge advantage that the safe data range can be modified after data reduction if needed.

**Comment by AtreyeeS** [comment]
I made the `livetime_map` hidden. We can expose this once we have some clarity on the meta handling. I can start to prepare a PIG for this. Maybe this PR can be merged independently ?

File: gammapy/makers/map.py (hunk 3 above)
Comment (AtreyeeS) [reply gammapy/makers/map.py]: Yes indeed, my bad. If the counts are there, then the livetime should also be.

File: gammapy/datasets/map.py [hunk 4]
```diff
@@ -127,6 +127,8 @@ class MapDataset(Dataset):
     meta_table : `~astropy.table.Table`
         Table listing information on observations used to create the dataset.
         One line per observation for stacked datasets.
+    livetime_map : `~gammapy.maps.WcsNDMap`
```
Comment (bkhelifi) [inline gammapy/datasets/map.py:130]: As the livetime_map is hidden, should we describe it into the docstring?

File: gammapy/datasets/map.py [hunk 5]
```diff
@@ -1877,6 +1887,7 @@ def resample_energy_axis(self, energy_axis, name=None):
         else:  # None or EDispMap
             kwargs["edisp"] = self.edisp
 
+        kwargs["_livetime_map"] = self._livetime_map
```
Comment (bkhelifi) [inline gammapy/datasets/map.py:1890]: Maybe one should add a `if self._livetime_map is not None:` ...

File: gammapy/datasets/map.py [hunk 6]
```diff
@@ -2239,7 +2255,7 @@ def to_map_dataset(self, name=None):
         The background model template is taken as alpha * counts_off
 
         Parameters
-        ----------
+        ----------c
```
Comment (bkhelifi) [inline gammapy/datasets/map.py:2258]:
```suggestion
        ----------
```

File: gammapy/datasets/tests/test_map.py [hunk 7]
```diff
@@ -930,6 +934,7 @@ def test_stack(sky_model):
     assert_allclose(stacked.mask_safe.data.sum(), 4600)
     assert_allclose(stacked.mask_fit.data.sum(), 4600)
     assert_allclose(stacked.exposure.data.sum(), 1.6e11)
+    assert_allclose(stacked._livetime_map.data.sum(), 240000.0)
 
     assert_allclose(stacked.meta_table["OBS_ID"][0], [0, 1])
```
Comment (bkhelifi) [inline gammapy/datasets/tests/test_map.py:940]: One could profit of this PR to check if the metadata livetime exists and contains the correct value

**Comment by bkhelifi** [comment]
A side question, do you test that the mask is correctly applied when computing the livetime_map?

**Comment by adonath** [comment]
The livetime map could be handled as meta information in its own data structure. See #4491.

**Comment by bkhelifi** [comment]
Based on the work associated to the PIG 25 about metadata container, this information could be added in the metadata of the dataset class.
What is the best strategy? Wait for this DatasetMetadataClass or do the merge and then update the code with the future metadata class?

**Comment by AtreyeeS** [comment]
Tutorial showed in #4817 
Meta information will be added later

**Pull request closed without merging**